ODIN_II/SRC/verilog_preprocessor.cpp

```diff
@@ -672,137 +671,161 @@ void push(veri_flag_stack *stack, int flag)
 	{
-		line[i++] = ch;
-		if (ch == '`')
+		directionDefined = true;
+		char * temp = NULL;
+		while(**subtoken != ' ')
+		{
+			postDec[(*j)++] = **subtoken;
+			(*subtoken)++;
+		}
+		postDec[(*j)++] = ' ';
+		(*subtoken)++;
+		std::string str(*subtoken);
+		// I/O type
+		if(str.find("reg[") == 0 || str.find("reg ") == 0 || str.find("wire[") == 0 || str.find("wire ") == 0)
 		{
-			while( (ch = getc(source)) != '\n')
+			temp = *subtoken;
+			do { 
+				(*subtoken)++;
+			} while (**subtoken != ' ' && **subtoken != '[');
+			if(**subtoken == ' ')
 			{
-				line[i++] = ch;
+				(*subtoken)++;
 			}
-			line[i] = '\n';
-			fputs(line,destination);
-			i = 0;
+			do {
+				IOTypeDec[(*k)++] = *temp;
+				temp++;
+			} while (*temp != '\0');
+			IOTypeDec[(*k)++] = ';';
+			IOTypeDec[(*k)++] = '\n';
+		}
+		// I/O size
+		if(**subtoken == '[')
+		{
+			while(**subtoken != ']')
+			{
+				postDec[(*j)++] = **subtoken;
+				(*subtoken)++;
+			}
+			postDec[(*j)++] = ']';
+			(*subtoken)++;
 		}
-
-	}
-	line[i++] = ch;
-	line[i] = '\0';
-	std::string sub_line = find_substring(line,"module",2);
-	sprintf(line,"%s",sub_line.c_str());
-	line = search_replace(line,"\n","",2);
-
-	if (! validate_string_regex(line, pattern))
-	{
-		free(line);
-	 	rewind(source);
-		return source;
 	}
-	for (i = 0; i < 4; i++)
+	if(**subtoken == ' ')
 	{
-		char *line_cpy = vtr::strdup(line);
-		free(line);
-		line = search_replace(line_cpy,searchString[i],"",2);
-		vtr::free(line_cpy);
+		(*subtoken)++;
 	}
-	i = 0;
-	while (i < strnlen(line, MaxLine))
+	// Variable name
+	while(**subtoken != NULL)
 	{
-		if(line[i] == '[')
+		if(directionDefined)
 		{
-			j = 0;
-			while(line[i] != ']')
-			{
-				temp[j++] = line[i];
-				i++;
-			}
-			temp[j++] = line[i];
-			temp[j] = '\0';
-			line = search_replace(line,temp,"",2);
-			i = 0;
+			postDec[(*j)++] = **subtoken;
 		}
-		else
-			i++;
+		dec[(*i)++] = **subtoken;
+		(*subtoken)++;
+	}
+	if(directionDefined)
+	{
+		postDec[(*j)++] = ';';
+		postDec[(*j)++] = '\n';
 	}
-	fputs(line, destination);
-	vtr::free(line);
-	rewind(source);
-	destination = format_verilog_variable(source,destination);
-	rewind(destination);
-	return destination;
 }
 
-FILE *format_verilog_variable(FILE * src, FILE *dest)
+FILE *format_verilog_file(FILE *source)
 {
-	char ch;
-	int i;
-	char line[MaxLine];
-	char *tempLine;
-	char *pos;
-	while( (ch = getc(src) ) != ';')
+	FILE *destination = tmpfile();
+	char buf[MaxLine] = { 0 }; // Temporary input buffer
+	char dec[MaxLine] = { 0 }; // Module declaration buffer
+	char postDec[MaxLine] = { 0 }; // Post-declaration buffer
+	char IOTypeDec[MaxLine] = { 0 }; // Register and wire re-declaration buffer
+	char * bufPtr = buf;
+	char * decPtr = dec;
+	char * postDecPtr = postDec;
+	char * IOTypeDecPtr = IOTypeDec;
+	char * token, * subtoken = NULL;
+	int pos = 0;
+	unsigned i, j, k = 0;
+
+	while(fgets(buf, MaxLine, source))
```

Comment:
I would trim here after fgets so that you don't have to do it manually.

Reply:
also, I would advise against having so many nested loops and fgets, only one fgets is clearer and avoids errors. Just change the behaviour based on what you are expecting. each behaviour can become its own small function and it makes it easier to grasp.